src/hooks/useLoader.ts

```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from "react";
+
+export function useLoader(): [
+  idle: boolean,
+  loading: boolean,
+  loaded: boolean,
+  setLoading: (val: boolean) => void
+] {
+  const [idle, setIdle] = useState(true);
+  const [requestsCount, setRequestsCounts] = useState(0);
+  const [loading, setLoading] = useState(false);
+  const [loaded, setLoaded] = useState(false);
+
+  useEffect(() => {
+    setLoading(requestsCount !== 0);
+    setLoaded(requestsCount === 0 && !idle);
+  }, [requestsCount]);
+
+  function computeLoadingState(showLoader: boolean) {
+    setIdle(false);
+    setRequestsCounts((val) => {
+      val = showLoader ? val + 1 : val - 1;
+      val = val < 0 ? 0 : val;
+      return val;
+    });
+  }
+  return [idle, loading, loaded, computeLoadingState];
+}
```

Comment:
Do we need tracking of multiple loading states? Logic would be simpler to reason if we are only tracking one. 

The complexity of multiple async ops is anyhow something we need to tackle at call site with promise cancellation etc. Keeping the useLoader straightforward would make it easier to reason. E.g. all it does is `idle → loading ↔ loaded`. No counting.

Also, instead of multiple variables, states could be represented in one variable. `state: ['loading',  'loaded', 'idle']`.

And `useLoader.ts` could export helper functions to check the state. call site e.g.

```js
import { useLoader, isLoading, isLoaded } from '../hooks/useLoader';

const { loadingState, setLoading } = useLoader();

const asyncFn = () {
  setLoading(true);
}

return (
<div>
  {isLoading(loadingState) && <Loading ... />}
  {isLoaded(loadingState) && <Data ... />}
</div>
)
```

Single state would be easier to understand and debug.